spikewrap/data_classes/base.py

```diff
@@ -0,0 +1,138 @@
+import fnmatch
+from collections import UserDict
+from collections.abc import ItemsView, KeysView, ValuesView
+from dataclasses import dataclass
+from itertools import chain
+from pathlib import Path
+from typing import Callable, Dict, List, Literal
+
+
+@dataclass
+class BaseUserDict(UserDict):
+    """
+    Base class for `PreprocessingData` and `SortingData`
+    used for checking and formatting `base_path`, `sub_name`
+    and `run_names`. The layout of the `rawdata` and
+    `derivatives` folder is identical up to the run
+    folder, allowing use of this class for
+    preprocessing and sorting.
+
+    Base UserDict that implements the
+    keys(), values() and items() convenience functions."""
+
+    base_path: Path
+    sub_name: str
+    sessions_and_runs: Dict
+
+    def __post_init__(self) -> None:
+        self.data: Dict = {}
+        self.base_path = Path(self.base_path)
+        self.check_run_names_are_formatted_as_list()
+
+    def check_run_names_are_formatted_as_list(self) -> None:
+        """"""
+        for key, value in self.sessions_and_runs.items():
+            if not isinstance(value, List):
+                assert isinstance(
+                    value, str
+                ), "Run names must be string or list of strings"
+                self.sessions_and_runs[key] = [value]
+
+    def preprocessing_sessions_and_runs(self):  # TODO: type hint
+        """"""
+        ordered_ses_names = list(
+            chain(*[[ses] * len(runs) for ses, runs in self.sessions_and_runs.items()])
+        )
+        ordered_run_names = list(
+            chain(*[runs for runs in self.sessions_and_runs.values()])
+        )
+
+        return list(zip(ordered_ses_names, ordered_run_names))
+
+    def _validate_inputs(
```

Comment:
I like this set of `assert` statements, seems like a clean way to hold all the validations together.
Do you think you will define separate methods to get the folders? For instance will the `get_sub_path` change depending on the implementation of `load_data()`?
In case yes it makes sense to inject the `Callable`s